crates/dyn-abi/src/eip712/parser.rs

```diff
@@ -139,13 +141,91 @@ impl<'a> EncodeType<'a> {
 
         Ok(Self { types })
     }
+
+    /// Computes the canonical string representation of the type.
+    ///
+    /// Orders the `ComponentTypes` based on the EIP-712 rules, and removes unsupported whitespaces.
+    pub fn canonicalize(&self) -> Result<String, Error> {
+        if self.types.is_empty() {
+            return Err(Error::MissingType("Primary Type".into()));
+        }
+
+        let primary_idx = self.get_primary_idx()?;
+
+        // EIP-712 requires alphabetical order of the secondary types
+        let mut types = self.types.clone();
+        let mut sorted = vec![types.remove(primary_idx)];
+        types.sort_by(|a, b| a.type_name.cmp(b.type_name));
+        sorted.extend(types);
+
+        // Ensure no unintended whitespaces
+        Ok(sorted.into_iter().map(|t| t.span.trim().replace(", ", ",")).collect())
+    }
+
+    /// Identifies the primary type from the list of component types.
+    ///
+    /// The primary type is the component type that is not used as a property in any component type
+    /// definition within this set.
+    fn get_primary_idx(&self) -> Result<usize, Error> {
+        // Track all defined component types and types used in component properties.
+        let mut components = HashSet::new();
+        let mut types_in_props = HashSet::new();
+
+        for ty in &self.types {
+            components.insert(ty.type_name);
+
+            for prop_def in &ty.props {
+                // Extract the base type name, removing array suffixes like "Person[]"
+                let type_str = prop_def.ty.span.trim();
+                let type_str = type_str.split('[').next().unwrap_or(type_str).trim();
+
+                // A type is considered a reference to another type if its name starts with an
+                // uppercase letter, otherwise it is assumed to be a basic type
+                if !type_str.is_empty()
+                    && type_str.chars().next().is_some_and(|c| c.is_ascii_uppercase())
+                {
+                    types_in_props.insert(type_str);
+                }
```

Comment:
I think the order can be swapped here because we only need to split at [ if this is uppercase?